src/lang/encoders.ml

```diff
@@ -0,0 +1,212 @@
+  let midi = kind_t ~pos ~level kind.Frame.midi in
+  format_t ~pos ~level (frame_kind_t ~pos ~level audio video midi)
+
+exception Error
+
+(* TODO: it would be better if each encoder registered itself here than
+   having this match *)
+let make_encoder_base e p =
+  let ogg_audio e p =
+    match e with
+      | "vorbis" -> Lang_vorbis.make p
+      | "vorbis.cbr" -> Lang_vorbis.make_cbr p
+      | "vorbis.abr" -> Lang_vorbis.make_abr p
+      | "opus" -> Lang_opus.make p
+      | "speex" -> Lang_speex.make p
+      | "flac" -> Lang_flac.make_ogg p
+      | _ -> raise Not_found
+  in
+  let ogg_audio_opt e p = try Some (ogg_audio e p) with Not_found -> None in
+  let ogg_video e p =
+    match e with "theora" -> Lang_theora.make p | _ -> raise Not_found
+  in
+  let ogg_video_opt e p = try Some (ogg_video e p) with Not_found -> None in
+  match e with
+    | "ogg" ->
+        let audio =
+          List.find_map
+            (function "", `Encoder (e, p) -> ogg_audio_opt e p | _ -> None)
+            p
+        in
+        let video =
+          List.find_map
+            (function "", `Encoder (e, p) -> ogg_video_opt e p | _ -> None)
+            p
+        in
+        Encoder.Ogg { Ogg_format.audio; video }
+    | "vorbis" ->
+        Encoder.Ogg
+          { Ogg_format.audio = Some (Lang_vorbis.make p); video = None }
+    | "vorbis.cbr" ->
+        Encoder.Ogg
+          { Ogg_format.audio = Some (Lang_vorbis.make_cbr p); video = None }
+    | "vorbis.abr" ->
+        Encoder.Ogg
+          { Ogg_format.audio = Some (Lang_vorbis.make_abr p); video = None }
+    | "opus" ->
+        Encoder.Ogg { Ogg_format.audio = Some (Lang_opus.make p); video = None }
+    | "speex" ->
+        Encoder.Ogg
+          { Ogg_format.audio = Some (Lang_speex.make p); video = None }
+    | "flac" -> Lang_flac.make p
+    | "theora" ->
+        Encoder.Ogg
+          { Ogg_format.audio = None; video = Some (Lang_theora.make p) }
+    | "ffmpeg" -> Lang_ffmpeg.make p
+    | "external" -> Lang_external_encoder.make p
+    | "gstreamer" -> Lang_gstreamer.make p
+    | "wav" -> Lang_wav.make p
+    | "avi" -> Lang_avi.make p
+    | "mp3" | "mp3.cbr" -> Lang_mp3.make_cbr p
+    | "mp3.abr" -> Lang_mp3.make_abr p
+    | "mp3.vbr" -> Lang_mp3.make_vbr p
+    | "mp3.fxp" | "shine" -> Lang_shine.make p
+    | "fdkaac" -> Lang_fdkaac.make p
+    | e -> (
+        try
+          Encoder.Ogg { Ogg_format.audio = Some (ogg_audio e p); video = None }
+        with Not_found -> (
+          try
+            Encoder.Ogg
+              { Ogg_format.audio = None; video = Some (ogg_video e p) }
+          with Not_found -> raise Error))
+
+let make_encoder ~pos t e p =
+  let e =
+    try make_encoder_base e p
+    with Error -> raise (Term.Unsupported_format (pos, Term.print_term t))
+  in
+  (try
+     let (_ : Encoder.factory) = Encoder.get_factory e in
+     ()
+   with Not_found -> raise (Term.Unsupported_format (pos, Term.print_term t)));
```

Comment:
Is that gonna raise to the runtime? Should we use a catchable runtime error instead?

Reply:
I am not sure here: most of the time we will check this at startup and get an error of the form
```
Error 12: Unsupported format: %bla().
You must be missing an optional dependency.
```
I think this is more clear than a catchable error, isn't it?